ui/src/app/utils/errors.tsx

```diff
@@ -85,3 +90,134 @@ export async function fetchWithSystemErrorHandler<T>(
     }
   }
 }
+
+// TODO handle errorClassName and parameters?
+export const defaultErrorResponseFormatter = (
+  errorResponse: ErrorResponse
+): string => {
+  const { message, statusCode, errorCode, errorUniqueId } = errorResponse;
+
+  const errorCodeStr =
+    errorCode && errorCode !== ErrorCode.PARSEERROR
+      ? ` of type ${errorCode.toString()}`
+      : '';
+  const messageStr = message ? `: ${message}` : '.';
+
+  const detailsStr = cond(
+    [
+      !!statusCode && !!errorUniqueId,
+      () =>
+        ` with HTTP status code ${statusCode} and unique ID ${errorUniqueId}`,
+    ],
+    [!!statusCode, () => ` with HTTP status code ${statusCode}`],
+    [!!errorUniqueId, () => ` with unique ID ${errorUniqueId}`],
+    () => ''
+  );
+
+  return `An API error${errorCodeStr} occurred${detailsStr}${messageStr}`;
+};
+
+export const FALLBACK_ERROR_TITLE = 'An error has occurred';
+
+/**
+ * Convert an API error response (of any type) to a format suitable for an error modal.  The caller may supply a
+ * matcher for expected error responses which should be considered successes, and a custom formatter for expected
+ * failure responses; if neither of these match, the defaultErrorResponseFormatter will be used.
+ *
+ * Example: an (async) API call getInfo() sometimes appropriately returns 404s which should not be considered errors,
+ * and there's a common error case (HTTP 429, Error Code TOO_FAST) which we want to handle specially.
+ *
+ * const responseP = getInfo();
+ * await responseP.then(resp =>
+ *  errorHandlerWithFallback(
+ *    resp,
+ *    (er) => er.statusCode === 404,
+ *    (er) =>
+ *      er.statusCode === 429 && er.errorCode === ErrorCode.TOO_FAST && {
+ *        title: 'Please try again',
+ *        message: 'The server is currently handling too many requests',
+ *      }
+ *    ));
+ *
+ * @param anyApiErrorResponse The error response from an API call, of any type
+ * @param expectedResponseMatcher An optional handler for errors which should be considered successes
+ * @param customErrorResponseFormatter An optional handler for expected responses; if missing or this handler returns
+ * undefined, the default formatter will be used.
+ * @returns A NotificationStore object suitable for an error modal, if appropriate; undefined otherwise
+ */
+export const errorHandlerWithFallback = (
+  anyApiErrorResponse: any,
+  expectedResponseMatcher?: (ErrorResponse) => boolean,
+  customErrorResponseFormatter?: (ErrorResponse) => NotificationStore
+): Promise<NotificationStore> =>
+  convertAPIError(anyApiErrorResponse).then((errorResponse) => {
```

Comment:
This function (`convertAPIError`) perhaps should be avoided. It takes an ErrorResponse, calls `.json()` on it, and then throws it away if that method fails.

Reply:
a new version of this is `toApiErrorResponse()` which retains all fields in the original and the `json()`